rfcs/45-stageless.md

```diff
@@ -0,0 +1,866 @@
+
+### The `App` stores multiple schedules
+
+The `App` can store multiple `Schedules`: each with their own `impl ScheduleLabel` type, allowing you to cleanly execute coherent blocks of logic when the need arises.
+By default, each app stores both a startup and main schedule: the former runs only once on app creation, while the latter loops endlessly.
+However, schedules are a surprisingly powerful and flexible tool: they can be used to handle initialization and cleanup logic when working with states or used ad-hoc to run game logic in complex patterns.
+
+The main, startup and rendering schedules can be accessed using the `DefaultSchedule::Main`, `DefaultSchedule::Startup`, `DefaultSchedule::Rendering` labels respectively.
+By default, systems are added to the main schedule.
+You can control this by adding the `.to_schedule(ScheduleLabel::Variant)` system descriptor to your system.
+To add an entirely new schedule to your app (which can be run on the world in an exclusive system), use `app.add_schedule(label, schedule)`.
+
+You can access the schedules stored in the app using the `&Schedules`.
+Unsurprisingly, this never conflicts with entities or resources in the `World`, as they are stored one level higher.
+Schedules can be modified by emitting `ScheduleCommands`, which apply to the specified schedule just before it is next started.
+For safety reasons, you cannot mutate schedules direclty: instead, you can defer their modification until just before next time that schedule is run using `ScheduleCommands`.
+
+```rust
+/// Each level has its own collection of dedicated systems
+/// in addition to the core logic
+#[derive(ScheduleLabel)]
+struct LevelSchedule(usize);
+
+struct CurrentLevel(0);
+struct PreviousLevel(0);
+
+impl LevelSchedule {
+   fn new(level: usize) -> Schedule {
+      // Insert diverse systems here based on which level was provided
+   }
+}
+
+fn main(){
+   App::new()
+      .add_plugins(DefaultPlugins)
+      .add_system(toggle_level_specific_systems)
+      .run()
+}
+
+/// Adds and removes systems from the schedule according to the currently active level
+///
+/// This could be done with run criteria instead, but this method will result in lower overhead
+fn toggle_level_specific_systems(current_level: Res<CurrentLevel>, previous_level: Res<PreviousLevel>, schedule_commands: ScheduleCommands){
+   if current_level.is_changed(){
+      // Add all of the systems from the current level's schedule to the main schedule, as if adding a plugin
+      schedule_commands.schedule(CoreSchedule::Main).merge_schedule(LevelSchedule::new(current_level.0))
+      // Remove all systems previously added by the old level's schedule
+      // Systems will only be removed if both the function signature and configuration are a perfect match.
+      schedule_commands.schedule(CoreSchedule::Main).extract_schedule(LevelSchedule::new(previous_level.0))

```

Comment:
> Systems will only be removed if both the function signature and configuration are a perfect match.

> If you wish to share behavior among siblings, add the systems repeatedly to each sibling, typically by saving a schedule and then using `Schedule::merge_schedule` to combine that into the specialized schedule of choice.

The topic of identifying systems in seems to be a bit loose.
Currently you can add as many systems with the same configuration as you want.
The RFC implies that the same signature and configuration guarantees uniqueness in the scope of a single schedule.
Meanwhile, in my opinion, function's signature should be enough to identify a system in most cases and the existence of 2 same ones could be a sign of user error.

Another thing worth pointing out is that merges are very ambiguous.
The example in states implies that adding multiple of the same system will result in only one existing.
It should be clearly decided which one gets discarded since their metadata could differ.

Again with the states example.
If we continue assuming that systems are also identified by their configuration
then the shared systems cannot in any way be used for ordering,
because they will be considered as different systems during a merge,
resulting in multiple of the same systems running.

Excuse me if those aren't correct or descriptive enough, I'm writing this at 7 AM.

Reply:
> Meanwhile, in my opinion, function's signature should be enough to identify a system in most cases and the existence of 2 same ones could be a sign of user error.

I don't think this is true. Consider index updating, or other synchronization / cleanup systems. There will be many copies of these in the schedule, and this is not a user error.

> The topic of identifying systems in seems to be a bit loose.

Strongly agree. I think that identifying systems by their signature + config is fundamentally doomed: we need to actually store which schedule they came from.

Reply:
> I don't think this is true. Consider index updating, or other synchronization / cleanup systems. There will be many copies of these in the schedule, and this is not a user error.

That's fair enough, `flush_commands` would be such a system as well, I did expect more to come.
Maybe, similarly to constraints, a system can decide how unique it has to be.
```rs
enum SystemRedundancy {
   Allow,
   WarnCount,
   WarnVerbose,
   Deny, 
   Forbid,
}
```
Pros:
 - warns when system is duplicated when it has to be original
 - can still have multi-use systems
Cons:
 - this should be stored per function-system signature, but that'd be more difficult
 - the system id is worse than composite, it can be either only signature or signature and config

Reply:
Interesting. I think it could be a useful additional correctness tool 🤔

Reply:
The schedule merging stuff feels pretty under-cooked. How would users resolve ambiguities that result from the merge? Is there an assumption that the merged schedules had labels in common?

Reply:
Yeah IMO we should just cut schedule merging. It's not MVP.

Reply:
Good call, it's complex enough for a small rfc